File: backend/src/main/java/br/com/guerreiro_dev/backend/domain/Locacao.java
package br.com.guerreiro_dev.backend.domain;

import br.com.guerreiro_dev.backend.domain.enums.StatusLocacao;
import jakarta.persistence.*;
import lombok.*;

import java.io.Serial;
import java.io.Serializable;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Getter
@Setter
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
@Entity
@NoArgsConstructor
@AllArgsConstructor
@Table(name = "tb_locacao")
public class Locacao implements Serializable {

    @Serial
    private static final long serialVersionUID = 1L;

    @EqualsAndHashCode.Include
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @ManyToOne
    @JoinColumn(name = "cliente_id")
    private Cliente cliente;

    @ManyToOne
    @JoinColumn(name = "veiculo_id")
    private Veiculo veiculo;

    private BigDecimal valorVeiculo;
    private BigDecimal valorTotal;
    private LocalDate dataRetirada;
    private LocalDate dataEntrega;

    @Enumerated(EnumType.STRING)
    private StatusLocacao status;

    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "tb_locacao_adicionais", joinColumns = @JoinColumn(name = "locacao_id"))
    private List<ItemAdicional> adicionais = new ArrayList<>();

    public void calcularValorTotal() {
        BigDecimal totalAdicionais = adicionais.stream()
                .map(ItemAdicional::getPreco)
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        this.valorTotal = valorVeiculo.add(totalAdicionais);
    }
}

File: frontend/types/locacao.ts
import { Cliente } from "./cliente";
import { Veiculo } from "./veiculo";

export interface ItemAdicional {
  nome: string;
  descricao: string;
  preco: number;
}

export type StatusLocacao =
  | "RESERVADO"
  | "EM_ANDAMENTO"
  | "FINALIZADO"
  | "CANCELADO";

export interface Locacao {
  id: string;

  cliente: Cliente;
  veiculo: Veiculo;

  valorVeiculo: number;
  valorTotal: number;

  dataRetirada: string;
  dataEntrega: string;

  status: StatusLocacao;

  adicionais: ItemAdicional[];
}

export type VeiculoCreate = Omit<Veiculo, 'id'>;